aten/src/ATen/native/cpu/UnaryOpsKernel.cpp

```diff
@@ -2,84 +2,164 @@
-static void parallel_apply(Tensor& result, const Tensor& self, F f) {
-  internal::init_tbb_num_threads();
-
-  static tbb::affinity_partitioner ap;
-
-  auto arr_out = result.data<scalar_t>();
-  auto arr_in = self.data<scalar_t>();
-  int64_t size = self.numel();
-  if (size < internal::TBB_GRAIN_SIZE) {
-    map(f, arr_out, arr_in, size);
-  } else {
-    tbb::parallel_for(
-        tbb::blocked_range<int64_t>(0, size, internal::TBB_GRAIN_SIZE),
-        [&](const tbb::blocked_range<int64_t>& r) {
-          map(f, arr_out + r.begin(), arr_in + r.begin(), r.end() - r.begin());
-        },
-        ap);
-  }
-}
-
 static void abs_kernel(Tensor& result, const Tensor& self) {
   AT_DISPATCH_ALL_TYPES(self.type(), "abs", [&] {
-    parallel_apply<scalar_t>(
+    CPU_tensor_parallel_kernel_apply2<scalar_t, scalar_t>(
         result,
         self,
-        [](const Vec256<scalar_t>& x) { return x.abs(); });  });
+        [](int64_t size,
+           scalar_t* x,
+           scalar_t* y,
+           int64_t stridex,
+           int64_t stridey) {
+          if (stridex == 1 && stridey == 1) {
+            map([](const Vec256<scalar_t>& x) { return x.abs(); }, x, y, size);
+          } else {
+            for (int64_t i = 0; i < size; i++) {
+              x[stridex * i] = std::abs(y[stridey * i]);
+            }
+          }
+        });
+  });
 }
 
-static void rsqrt_kernel(Tensor& result, const Tensor& self) {
-  AT_DISPATCH_FLOATING_TYPES(self.type(), "rsqrt", [&] {
-    parallel_apply<scalar_t>(
-        result,
-        self,
-        [](const Vec256<scalar_t>& x) { return Vec256<scalar_t>((scalar_t)(1)) / x.sqrt(); });  });
-}
+// [Note AVX-SSE transitions] In general we avoid calls into cmath for code
+// compiled with AVX/AVX2 This is because of SSE-AVX transitions and a bug in
+// Glibc2.23 See https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1663280
+// Calling zeroupper when using AVX/AVX2 code resolves this.
+#if defined(__AVX__) && defined(__GLIBC__) && __GLIBC_MINOR__ == 23
+#define ZEROUPPER _mm256_zeroupper();
+#else
+#define ZEROUPPER
+#endif
 
-#define IMPLEMENT_FLOAT_KERNEL(op)                                             \
-  static void op##_kernel(Tensor& result, const Tensor& self) {                \
-    AT_DISPATCH_FLOATING_TYPES(self.type(), #op, [&] {                         \
-      parallel_apply<scalar_t>(                                                \
-          result, self, [](const Vec256<scalar_t>& x) { return x.op(); }); \
-    });                                                                        \
-  }                                                                            \
+#define IMPLEMENT_FLOAT_COMPUTEBOUND_KERNEL(op, opfn)                   \
+  static void op##_kernel(Tensor& result, const Tensor& self) {         \
+    AT_DISPATCH_FLOATING_TYPES(self.type(), #op, [&] {                  \
+      static constexpr int WIDTH = 128 / sizeof(scalar_t);              \
```

Comment:
Is 128B really the sweet spot for buffering the input? What if you increase it to e.g. 512B? Ideally we would be filling up a large part of the cache.

Reply:
It's not clear what to do here. 128 was chosen since it's 2 cachelines. I've yet to experiment with this more. It's also not easy to make this work on different cache sizes then and the amount of cache on my dev machine is pretty unrepresentative of an average user.